### PR TITLE
Linear nodes attempt (v2)

### PR DESCRIPTION
**Pull request by juanitorduz**
Attempts to address https://github.com/pymc-devs/pymc-bart/issues/51

From the very initial attempt (https://github.com/pymc-devs/pymc-bart/pull/74)

> I really want to have the linear nodes hehe. So this is a draft PR to try to port the changes added in https://github.com/pymc-devs/pymc/pull/5044 . The code-base has changed a bit so I need to catch up in understanding the code more. If someone could give me some guidance it would be much appreciated. I feel this is a relatively "low-hanging-fruit" but as always, the devil is in the details.

For this iteration I suggest I add some code, add questions and incorporate feedback bit-by-bit.

### CODE REVIEW

File: pymc_bart/tree.py [hunk 1]
```diff
@@ -231,9 +234,14 @@ def _traverse_tree(
...
         """
-        current_node: Node = self.get_node(node_index)
+        current_node = self.get_node(node_index)
+
         if current_node.is_leaf_node():
-            return np.array(current_node.value)
+            if current_node.linear_params is not None:
+                return np.array(current_node.value)
+            x = x[split_variable].item()
+            y_x = current_node.linear_params[0] + current_node.linear_params[1] * x
```
Comment (juanitorduz) [inline pymc_bart/tree.py:243]: In which method should we incorporate this logic? Also, where should we put the variables `m` and `split_variable` (what is the type of `split_variable`?)

Comment (aloctavodia) [reply pymc_bart/tree.py:243]: here https://github.com/pymc-devs/pymc-bart/blob/main/pymc_bart/tree.py#L259-L260

split_variable is of type integer the name of that variable in the new code is  `idx_split_variable`.

we should pass `m` to `predict` (predict is called from _sample_posterior which in turn is called from plot_variable_importance and plot_dependence ). Assuming  `μ` is the BART RV we can get `m` from `μ.owner.op.m`

Comment (juanitorduz) [reply pymc_bart/tree.py]: In https://github.com/pymc-devs/pymc-bart/pull/79/commits/8f5b293e7da723d831dc35c3dae7b8d063d07bd4, https://github.com/pymc-devs/pymc-bart/pull/79/commits/b31bbd0731339e92b10a4947b47fc15ff130f8cf and https://github.com/pymc-devs/pymc-bart/pull/79/commits/8cf70289dd874f3b61ddaff2897362f09a331fc6 I tried to add the signatures of the functions and the changes in the `tree.py` module. You suggested adding some changes to ` _traverse_leaf_values`, however as this function does not have an `x` argument, I thought adding the changes in `_traverse_tree` directly. I am of course open for feedback :) . For now, mypy and the tests are happy :)

**Comment by juanitorduz** [comment]
I think we are (hopefully) ready to add the chances into the `pgbart.by` module. In https://github.com/pymc-devs/pymc-bart/pull/79/commits/7d3edc7b5ac9d8ef2996687a8b23ad3ad281f6bf I started with the basic signatures.  From the initial PR it looks we need to add something like:

```python
    if response == "mix":
        response = "linear" if np.random.random() >= 0.5 else "constant"

    left_node_value, left_node_linear_params = draw_leaf_value(
        sum_trees_output[left_node_idx_data_points],
        X[left_node_idx_data_points, selected_predictor],
        mean,
        linear_fit,
        m,
        normal,
        mu_std,
        response,
    )

    right_node_value, right_node_linear_params = draw_leaf_value(
        sum_trees_output[right_node_idx_data_points],
        X[right_node_idx_data_points, selected_predictor],
        mean,
        linear_fit,
        m,
        normal,
        mu_std,
        response,
    )
```

where should we add this? Or is there a recommended way of improving this?

**Comment by juanitorduz** [comment]
I squashed the past commits during the rebase 🤦 (sorry!)

File: pymc_bart/pgbart.py [hunk 2]
```diff
@@ -597,3 +609,27 @@ def logp(point, out_vars, vars, shared):  # pylint: disable=redefined-builtin
...
+        y_fit = a + b * X
+        return y_fit, (a, b)
+
+    try:
+        from numba import jit
+
+        return jit(linear_fit)
+
+    except ImportError:
+        return linear_fit
```
Comment (aloctavodia) [inline pymc_bart/pgbart.py:635]: numba is a dependency, so just do this. Also move this up closer to `fast_mean`

```suggestion
def fast_linear_fit(X, Y):
    """Use Numba to speed up the computation of the linear fit"""
    n = len(Y)
    xbar = np.sum(X) / n
    ybar = np.sum(Y) / n

    b = (X @ Y - n * xbar * ybar) / (X @ X - n * xbar**2)
    a = ybar - b * xbar

    y_fit = a + b * X
    return y_fit, (a, b)
```

File: pymc_bart/pgbart.py [hunk 3]
```diff
@@ -160,6 +165,9 @@ def __init__(
             num_observations=self.num_observations,
             shape=self.shape,
         )
+
+        self.linear_fit = fast_linear_fit()
```
Comment (aloctavodia) [inline pymc_bart/pgbart.py:169]: No need to define this here (and then pass it to the different functions) just call it  inside `draw_leaf_value`

Comment (juanitorduz) [reply pymc_bart/pgbart.py]: removed it https://github.com/pymc-devs/pymc-bart/pull/79/commits/613d0ecc6b65a69dcdf935a562f8ef5d68e385c7

**Comment by juanitorduz** [comment]
Thank you @aloctavodia for the quick feedback! I think `mypy` is happy now (all matches up!) However, I am getting weird errors running the tests 

```python
pymc.sampling.parallel.RemoteTraceback: 
"""
Traceback (most recent call last):
  File "/Users/juanitorduz/.pyenv/versions/3.10.10/lib/python3.10/site-packages/pymc/sampling/parallel.py", line 122, in run
    self._start_loop()
  File "/Users/juanitorduz/.pyenv/versions/3.10.10/lib/python3.10/site-packages/pymc/sampling/parallel.py", line 174, in _start_loop
    point, stats = self._step_method.step(self._point)
  File "/Users/juanitorduz/.pyenv/versions/3.10.10/lib/python3.10/site-packages/pymc/step_methods/compound.py", line 231, in step
    point, sts = method.step(point)
  File "/Users/juanitorduz/.pyenv/versions/3.10.10/lib/python3.10/site-packages/pymc/step_methods/arraystep.py", line 100, in step
    apoint, stats = self.astep(q)
  File "/Users/juanitorduz/Documents/pymc-bart/pymc_bart/pgbart.py", line 208, in astep
    if p.sample_tree(
  File "/Users/juanitorduz/Documents/pymc-bart/pymc_bart/pgbart.py", line 68, in sample_tree
    idx_new_nodes = grow_tree(
  File "/Users/juanitorduz/Documents/pymc-bart/pymc_bart/pgbart.py", line 426, in grow_tree
    node_value, linear_parms = draw_leaf_value(
  File "/Users/juanitorduz/.pyenv/versions/3.10.10/lib/python3.10/site-packages/numba/core/dispatcher.py", line 468, in _compile_for_args
    error_rewrite(e, 'typing')
  File "/Users/juanitorduz/.pyenv/versions/3.10.10/lib/python3.10/site-packages/numba/core/dispatcher.py", line 409, in error_rewrite
    raise e.with_traceback(None)
numba.core.errors.TypingError: Failed in nopython mode pipeline (step: nopython frontend)
Cannot unify float64 and array(float64, 1d, C) for 'mu_mean.4', defined at /Users/juanitorduz/Documents/pymc-bart/pymc_bart/pgbart.py (484)

File "pymc_bart/pgbart.py", line 484:
def draw_leaf_value(y_mu_pred, x_mu, m, norm, shape, response):
    <source elided>

    return norm + mu_mean, linear_params
```

Which I think it boils down to a `numba` error:
```python
numba.core.errors.TypingError: Failed in nopython mode pipeline (step: nopython frontend)
```
probably something with the `fast_linear_fit` function? Any tip would be welcome!

**Comment by juanitorduz** [comment]
@aloctavodia In https://github.com/pymc-devs/pymc-bart/pull/79/commits/796c9a69ffd001187abb7fa2a9b6096798c33e63 I fixed the numba shapes 🙌 ! The CI passes (current tests and mypy). 
In https://github.com/pymc-devs/pymc-bart/pull/79/commits/f4e7074eb9ee74ffce5e5cd3600333bc12a8d1ef I added a test to test the linear response and it is currently failing since I think there is a problem with the vectorization of the linear model.

File: pymc_bart/pgbart.py [hunk 4]
```diff
@@ -456,17 +465,23 @@ def get_split_value(available_splitting_values, idx_data_points, missing_data):
...
 
     if y_mu_pred.size == 1:
         mu_mean = np.full(shape, y_mu_pred.item() / m)
     else:
-        mu_mean = fast_mean(y_mu_pred) / m
-
-    return norm + mu_mean
+        if response == "linear":
+            y_fit, linear_params = fast_linear_fit(x=x_mu, y=y_mu_pred)
+            mu_mean = fast_mean(y_fit.reshape(1, -1)) / m
```
Comment (juanitorduz) [inline pymc_bart/pgbart.py:480]: The model runs in for the linear response in the scalar case (see tests). However, it fails on the test with non-trivial shapes. I am actually not convinced with this piece of code to define the output. Do you have any suggestions?

**Comment by juanitorduz** [comment]
> @aloctavodia In https://github.com/pymc-devs/pymc-bart/pull/79/commits/796c9a69ffd001187abb7fa2a9b6096798c33e63 I fixed the numba shapes 🙌 ! The CI passes (current tests and mypy). 
> 
> In https://github.com/pymc-devs/pymc-bart/pull/79/commits/f4e7074eb9ee74ffce5e5cd3600333bc12a8d1ef I added a test to test the linear response and it is currently failing since I think there is a problem with the vectorization of the linear model.

Now the base implementation with linear response seems to work. The issue now is the vectorization for  shape parameters 🥲

File: pymc_bart/pgbart.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (juanitorduz) [reply pymc_bart/pgbart.py]: @aloctavodia Now I remember why I though about reducing with the mean: because otherwise numba does not compile because a problems in dimensions (see ci)
```python
Cannot unify array(float64, 2d, C) and array(float64, 1d, C) for 'mu_mean.4', defined at /home/runner/work/pymc-bart/pymc-bart/pymc_bart/pgbart.py (483)
```
This is what has been driving me crazy 😅

[the remark this answers is not in the document]
Comment (aloctavodia) [reply pymc_bart/pgbart.py]: For the moment we can just comment out `@njitt` for that function. Not sure how much numba is helping with that function anyways. But if we still want numba we may need to refactor `draw_leaf_values` into one or more jitted functions with a well-defined output and one unfitted function with a variable output.

**Review by aloctavodia — COMMENTED** [review]
The copy method is missing one value `linear_params`

    def copy(self) -> "Tree":
        tree: Dict[int, Node] = {
            k: Node(v.value, v.idx_data_points, v.idx_split_variable, v.linear_params)
            for k, v in self.tree_structure.items()
        }
        idx_leaf_nodes = self.idx_leaf_nodes.copy() if self.idx_leaf_nodes is not None else None
        return Tree(tree_structure=tree, idx_leaf_nodes=idx_leaf_nodes, output=self.output)

File: pymc_bart/bart.py [hunk 5]
```diff
@@ -88,6 +91,7 @@ def __new__(
         Y: TensorLike,
         m: int = 50,
         alpha: float = 0.25,
+        response: str = "mix",
```
Comment (aloctavodia) [inline pymc_bart/bart.py:94]: Currently, there is no logic to handle this. So we only have linear or constant right? I think this is fine, let's keep only those two options, at least for this PR.

File: pymc_bart/pgbart.py [hunk 6]
```diff
@@ -488,6 +503,20 @@ def fast_mean(ari):
...
+    """Use Numba to speed up the computation of the linear fit"""
+    n = len(x)
+    xbar = np.sum(x) / n
+    ybar = np.sum(y) / n
+
+    b = (x @ y.T - n * xbar * ybar) / (x @ x - n * xbar**2)
+    a = ybar - b * xbar
+
+    y_fit = a + b * x
+    return y_fit, [a, b]
```
Comment (aloctavodia) [inline pymc_bart/pgbart.py:517]: Not sure about this, but probably we want

```suggestion
    return y_fit, [a.item(), b.item()]
```

File: pymc_bart/utils.py [hunk 7]
```diff
@@ -57,7 +60,7 @@ def _sample_posterior(
         flatten_size *= s
 
     idx = rng.integers(0, len(stacked_trees), size=flatten_size)
-    shape = stacked_trees[0][0].predict(X[0]).size
+    shape = stacked_trees[0][0].predict(x=X[0], m=m).size
```
Comment (aloctavodia) [inline pymc_bart/utils.py:63]: This actually for a line below this

            p += np.vstack([tree.predict(x, excluded) for x in X])

should be
            
             p += np.vstack([tree.predict(x, m=m, excluded=excluded) for x in X])

Comment (juanitorduz) [reply pymc_bart/utils.py:63]: added in https://github.com/pymc-devs/pymc-bart/pull/79/commits/996596f2053bd4bb1b942a5b483d712d84b042c0

File: pymc_bart/pgbart.py (hunk 6 above)
Comment (juanitorduz) [reply pymc_bart/pgbart.py]: added in https://github.com/pymc-devs/pymc-bart/pull/79/commits/996596f2053bd4bb1b942a5b483d712d84b042c0

File: pymc_bart/bart.py (hunk 5 above)
Comment (juanitorduz) [reply pymc_bart/bart.py:94]: added in https://github.com/pymc-devs/pymc-bart/pull/79/commits/996596f2053bd4bb1b942a5b483d712d84b042c0 as in the original PR

File: pymc_bart/pgbart.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (juanitorduz) [reply pymc_bart/pgbart.py]: In https://github.com/pymc-devs/pymc-bart/pull/79/commits/996596f2053bd4bb1b942a5b483d712d84b042c0 I changed the linear fir formula to 
```python
b = ((x - xbar) @ (y - ybar).T) / ((x - xbar) @ (x - xbar).T)
```
and `numba` stop complaining :)

**Comment by juanitorduz** [comment]
ok! I think the code looks much nice now and I believe we are getting closer bit-by-bit 😅 ! The `draw_leaf_value` is clean and compiles to `numba` via `njit`. However, we are getting a weird shape error which I have not been able to fix afters hours of debugging. 

```python
import pymc_bart as pmb
import pymc as pm
import numpy as np

X = np.random.normal(0, 1, size=(250, 3))
Y = np.random.normal(0, 1, size=250)
X[:, 0] = np.random.normal(Y, 0.1)

with pm.Model() as model:
    mu = pmb.BART("mu", X, Y, m=10, response="linear")
    sigma = pm.HalfNormal("sigma", 1)
    y = pm.Normal("y", mu, sigma, observed=Y)
    idata = pm.sample(random_seed=3415, chains=2)
```

yields

```python
ValueError                                Traceback (most recent call last)
File [~/opt/anaconda3/envs/pymc-bart-env/lib/python3.10/site-packages/pytensor/link/vm.py:414](https://file+.vscode-resource.vscode-cdn.net/Users/juanitorduz/Documents/pymc-bart/~/opt/anaconda3/envs/pymc-bart-env/lib/python3.10/site-packages/pytensor/link/vm.py:414), in Loop.__call__(self)
    411 for thunk, node, old_storage in zip_longest(
    412     self.thunks, self.nodes, self.post_thunk_clear, fillvalue=()
    413 ):
--> 414     thunk()
    415     for old_s in old_storage:

File [~/opt/anaconda3/envs/pymc-bart-env/lib/python3.10/site-packages/pytensor/graph/op.py:543](https://file+.vscode-resource.vscode-cdn.net/Users/juanitorduz/Documents/pymc-bart/~/opt/anaconda3/envs/pymc-bart-env/lib/python3.10/site-packages/pytensor/graph/op.py:543), in Op.make_py_thunk..rval(p, i, o, n, params)
    539 @is_thunk_type
    540 def rval(
    541     p=p, i=node_input_storage, o=node_output_storage, n=node, params=None
    542 ):
--> 543     r = p(n, [x[0] for x in i], o)
    544     for o in node.outputs:

File [~/opt/anaconda3/envs/pymc-bart-env/lib/python3.10/site-packages/pytensor/tensor/random/op.py:378](https://file+.vscode-resource.vscode-cdn.net/Users/juanitorduz/Documents/pymc-bart/~/opt/anaconda3/envs/pymc-bart-env/lib/python3.10/site-packages/pytensor/tensor/random/op.py:378), in RandomVariable.perform(self, node, inputs, outputs)
    376 rng_var_out[0] = rng
--> 378 smpl_val = self.rng_fn(rng, *(args + [size]))
    380 if (
    381     not isinstance(smpl_val, np.ndarray)
    382     or str(smpl_val.dtype) != out_var.type.dtype
    383 ):

File [~/Documents/pymc-bart/pymc_bart/bart.py:55](https://file+.vscode-resource.vscode-cdn.net/Users/juanitorduz/Documents/pymc-bart/~/Documents/pymc-bart/pymc_bart/bart.py:55), in BARTRV.rng_fn(cls, rng, X, Y, m, alpha, split_prior, size)
     54 else:
---> 55     return _sample_posterior(cls.all_trees, cls.X, cls.m, rng=rng).squeeze().T

File [~/Documents/pymc-bart/pymc_bart/utils.py:69](https://file+.vscode-resource.vscode-cdn.net/Users/juanitorduz/Documents/pymc-bart/~/Documents/pymc-bart/pymc_bart/utils.py:69), in _sample_posterior(all_trees, X, m, rng, size, excluded)
     68     for tree in stacked_trees[idx[ind]]:
---> 69         p += np.vstack([tree.predict(x=x, m=m, excluded=excluded) for x in X])
     70 pred.reshape((*size_iter, shape, -1))

File <__array_function__ internals>:200, in vstack(*args, **kwargs)

File [~/opt/anaconda3/envs/pymc-bart-env/lib/python3.10/site-packages/numpy/core/shape_base.py:296](https://file+.vscode-resource.vscode-cdn.net/Users/juanitorduz/Documents/pymc-bart/~/opt/anaconda3/envs/pymc-bart-env/lib/python3.10/site-packages/numpy/core/shape_base.py:296), in vstack(tup, dtype, casting)
    295     arrs = [arrs]
--> 296 return _nx.concatenate(arrs, 0, dtype=dtype, casting=casting)

File <__array_function__ internals>:200, in concatenate(*args, **kwargs)

ValueError: all the input array dimensions except for the concatenation axis must match exactly, but along dimension 1, the array at index 0 has size 104 and the array at index 1 has size 143

During handling of the above exception, another exception occurred:
...
Inputs values: [Generator(PCG64) at 0x7FA8D903A7A0, array([], dtype=int64), array(11), 'not shown', 'not shown', array(10, dtype=int8), array(0.25, dtype=float32), array([], dtype=float64)]
Outputs clients: [['output'], [Elemwise{second,no_inplace}(mu, TensorConstant{(1,) of -0..5211646611})]]
```

which is a shape error 😢 . 

Any tips are welcome!

**Review by aloctavodia — COMMENTED** [review]
We forgot to add "v.linear_params" to the trim function. Something like this

            k: Node(v.value, None, v.idx_split_variable, v.linear_params) for k, v in self.tree_structure.items()
            
The trim function is used to delete some attributes we no longer need once we have sampled a tree, we do this just to save memory.

**Comment by juanitorduz** [comment]
Thanks! Fixed in https://github.com/pymc-devs/pymc-bart/pull/79/commits/6d4678423b2d7827d5af1289f93e21e6c2906d67

**Comment by juanitorduz** [comment]
@aloctavodia thanks for the suggestion! In https://github.com/pymc-devs/pymc-bart/pull/79/commits/c2de9fb3acef3e611539bb5b2ec2bc1f393a279e I added your code. The tests for `tests/test_bart.py::test_shape` keep failing. Surprisingly, the simple unit tests `tests/test_pgbart.py::test_fast_linear_fit` started falling. I looked into it and apparently, numba does not like the `if ... else`  condition. Hence, I took I simplified it in https://github.com/pymc-devs/pymc-bart/pull/79/commits/8c6963d27a063836273bcdb91abbf433771598f2 and these tests pass now (so maybe the `b = np.zeros(1)` was not necessary?) . 

Therefore I think we still have two problems for the linear response:
- `tests/test_bart.py::test_shape`
- `tests/test_bart.py::test_missing_data` sometimes fail and sometimes passes (see tests of https://github.com/pymc-devs/pymc-bart/pull/79/commits/8c6963d27a063836273bcdb91abbf433771598f2 for different python versions)

**Comment by juanitorduz** [comment]
The vectorization has been resolved 🥳 ! The only edge case missing is `tests/test_bart.py::test_missing_data[linear_response]`

File: pymc_bart/pgbart.py [hunk 8]
```diff
@@ -415,18 +421,24 @@ def grow_tree(
...
     for idx in range(2):
         idx_data_point = new_idx_data_points[idx]
-        node_value = draw_leaf_value(
-            sum_trees[:, idx_data_point],
-            m,
-            normal.rvs() * kfactor,
-            shape,
+        node_value, linear_params = draw_leaf_value(
+            y_mu_pred=sum_trees[:, idx_data_point],
+            x_mu=X[idx_data_point, selected_predictor],
```
Comment (juanitorduz) [inline pymc_bart/pgbart.py:431]: @aloctavodia  I think the problem is here as we are not handling the missing values for the linear model anywhere. A brute force approach adding `np.nan_to_num(X[idx_data_point, selected_predictor])` (see [numpy.nan_to_num](https://numpy.org/doc/stable/reference/generated/numpy.nan_to_num.html)) makes the model run. We could fill in missing values or simply warn the user missing values with `linear` response (and therefore `response = mix`) is not supported? 

I do not see anywhere in the PR in the PyMC repo any type of imputation.

**Comment by juanitorduz** [comment]
I think is finally ready for review 😆 !

File: pymc_bart/pgbart.py (hunk 8 above)
Comment (juanitorduz) [reply pymc_bart/pgbart.py:433]: Thanks for the tip, I resolved thin in https://github.com/pymc-devs/pymc-bart/pull/79/commits/c24262803a5d443145bf423502b7ec2cfa434c1e

File: pymc_bart/pgbart.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (juanitorduz) [reply pymc_bart/pgbart.py]: solved in https://github.com/pymc-devs/pymc-bart/pull/79/commits/f2c546a33059fd8f96375f5fb8a81b2f364a7fc0

Comment (juanitorduz) [reply pymc_bart/pgbart.py]: solved in https://github.com/pymc-devs/pymc-bart/pull/79/commits/f2c546a33059fd8f96375f5fb8a81b2f364a7fc0

**Comment by aloctavodia** [comment]
thanks a lot for this @juanitorduz!!!
